### PR TITLE
fix(select): keep ngModel when selected option is recreated with ngRe…

### LINKED ISSUE #15630: Angular 1.6 incorrectly resets `<select>'s `ng-model` value
***Note*: for support questions, please use one of these channels: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#question. This repository's issues are reserved for feature requests and bug reports.**

**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
Angular 1.6 resets `<select>`s `ng-model` value when `<option>`s list changes (see "Other information" section).
Angular 1.5 preserves it.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via https://plnkr.co or similar (template: http://plnkr.co/edit/tpl:yBpEi4).**
Angular 1.6: https://plnkr.co/edit/wHc4rzy7x9PAEqQLfVQT?p=preview
Angular 1.5.11: https://plnkr.co/edit/e0LVMEtGP2j4JxNf8lsp?p=preview

**What is the expected behavior?**
`ng-model` preserves it's value.

**What is the motivation / use case for changing the behavior?**
It's not correct.

**Which versions of Angular, and which browser / OS are affected by this issue? Did this work in previous versions of Angular? Please also test with the latest stable and snapshot (https://code.angularjs.org/snapshot/) versions.**
Angular 1.6.x is affected.
Angular 1.5.x is not.

**Other information (e.g. stacktraces, related issues, suggestions how to fix)**
The thing is Angular 1.6 tracks options by it's model and not by the `options`s value.

In the snippets above options array contains 2 objects: `[{val: "1"}, {val: "2"}]`.
`ng-repeat` iterates through them and uses `val` property as both option value and title.
`vm.selected` is used as select's `ng-model` value and is initially set to "2".

After pressing `Change options` button I replace options array with this one: `[{val: "2"}, {val: "3"}]` and Angular 1.6 resets `vm.selected` value **but it shouldn't** because option with value `2` is still in the list.

Angular 1.5 behaves correctly in this case and `vm.selected` value is preserved.

If `vm.options` contains primitive values bug doesn't appear ([snippet](https://plnkr.co/edit/tSd6MGChGB4s0eX1UIVS?p=preview)).

### PR DESCRIPTION
**Pull request by Narretz**
…peat

Fixes #15630

### CODE REVIEW

**Comment by Narretz** [comment]
Looks like my test passes without the patch 😨 Let's see

**Comment by Narretz** [comment]
Test has been fixed

File: test/ng/directive/selectSpec.js [hunk 1]
```diff
@@ -2316,6 +2316,36 @@ describe('select', function() {
...
+          );
+
+          var optionElements = element.find('option');
+          expect(optionElements.length).toEqual(3);
+          expect(optionElements[0].value).toBe('A');
+          expect(optionElements[1]).toBeMarkedAsSelected();
+          expect(scope.obj.value).toBe('B');
+
+          scope.$apply(function() {
+            // Only when new objects are used, ngRepeat re-creates the element from scratch
```
Comment (gkalpak) [inline test/ng/directive/selectSpec.js]: Just to make it more explicit, I would add an expactation that `optionElements[1] (before) !== optionElements[0] (after)`.

**Review by gkalpak — APPROVED** [review]
LGTM (with one optional suggestion).

**Comment by Narretz** [comment]
Updated, PTAL

File: test/ng/directive/selectSpec.js [hunk 2]
```diff
@@ -2316,6 +2316,40 @@ describe('select', function() {
...
+            scope.options = [{ name: 'B'}, { name: 'C'}, { name: 'D'}];
+          });
+
+          var previouslySelectedOptionElement = optionElements[1];
+          optionElements = element.find('option');
+          expect(optionElements.length).toEqual(3);
+
```
Comment (gkalpak) [inline test/ng/directive/selectSpec.js]: Misplaced newline?

**Review by gkalpak — APPROVED** [review]

**Comment by odedniv** [comment]
This is a really bad breakage, why isn't this in 1.6 already? Is it waiting for something I can help with?

**Comment by Narretz** [comment]
Hi @odedniv this is in 1.6 as https://github.com/angular/angular.js/commit/131af8272d269a541d04cb522c264a91e0ec8b6a, released with 1.6.2
Not sure why you think it's not in 1.6 yet

**Comment by odedniv** [comment]
@Narretz right, wasn't looking properly, but I'm still having issues with `<select ng-model>` resetting value that I don't have in 1.5. Any ideas?

**Comment by Narretz** [comment]
@odedniv I don't know of any other issues and since my crystal ball is still in repair, I can't say what might be going on with your selects. Please open a new issue with a minimal reproduction of the problem. Thanks!

**Pull request merged**